### PR TITLE
feat: add iroh-sync and integrate into iroh node

### PR DESCRIPTION
**Pull request by Frando**
## Description

This PR adds `iroh-sync`, a document synchronization protocol, to iroh, and integrates with `iroh-net`, `iroh-gossip` and `iroh-bytes`. 

* At the core is the `iroh-sync` crate, with a set reconciliation algorithm implemented by @dignifiedquire. See [the old iroh-sync repo](https://github.com/n0-computer/iroh-sync/) for the prehistory and #1216 for the initial PR (fully included in this PR, and by now outdated)
* Iroh sync is integrated in the iroh node, with iroh-gossip, in the RPC interface, and the CLI. 
  * `LiveSync` is the handle to an actor that integrates sync with [gossip](#1149 ) to broadcast and receive document updates from peers. For each open document a gossip swarm is joined with a `TopicId` derived from the doc namespace.
  * mod `download` contains the new downloader. It will be improved in #1344 .
  *  mod `client` is the new high-level RPC client. It currently only has methods for dealing with docs and sync, other things should be added once we merged this. CLI commands for sync are in `commands/sync.rs`. Will be much better with #1356 .
  *  `examples/sync.rs`  has a REPL to modify and sync docs. It does a full setup without using the iroh console. Also includes code to sync directories, and a hammer command for load testing.
* The PR also introduces `iroh::client::Iroh`, a wrapper around the RPC client, and   `iroh::client::Doc`, a wrapper around RPC client for a single document

#### Should likely happen before merge:

* [x] Make `iroh_sync::Store:::list_authors` and `list_replicas` return iterators `iroh-sync` *fixed in #1366 *
* [ ] Add `iroh_sync::Store::close_replica`
* [x] `ContentStatus` in `on_insert` callback is reported as `Ready` if the content is still `baomap::PartialEntry` (in-process download) *fixed in a8e809367245504787168d354f1714b11305920d*

#### Can happen after merge, but  before `0.6` release

* [ ] Implement `AuthorImport` and `AuthorShare` RPC & CLI commands
* [ ] sync store `list_namespaces` and `list_authors` internally collect, return iterator instead
* [ ] Fix cross-compiles to arm/android. See https://github.com/cross-rs/cross/issues/1311
* [ ] Ensure that fingerpring calculation is efficient and/or cached for large documents. Currently calculating the initial fingerprint iterates once over all entries in a document.
* [ ] Make content downloads be more reliable
   * [ ] Add some way to download content from peers independent of the first insertion event for a remote entry. The downloader with retries is tracked in #1334 and 1344, but independent of that, we still would currently only ever try to queue a download when the `on_insert` callback triggers, which is only once. There should be a way, even if manual for now, to try to download missing content in a replica from peers.
   * [ ] during `iroh-sync` sync include info if content is available for each entry
* [ ] Add basic peer management and persistence. Currently live sync will stop to do anything after a node restart.
    * [ ] Persist the addressbook of peers for a document, to reconnect when restarting the node
    * [ ] Implement `PeerAdd` and `PeerList` RPC & CLI commands. The latter needs changes in `iroh-net` to expose information of currently-connected peers and their peer info.
* [ ] Make read-only replicas possible
* [ ] Improve reliablity of replica sync. 
     * sync is triggered on each `NeighborUp` event from gossip. check that we don't sync too much.
     * maybe include peer info in gossip messages, to queue syncs with those (but not all at once)
     * track and exchange the timestamp of last full sync for peers, to know if you missed gossiped message and react accordingly
     * add more tests with peers coming and leaving

#### Open questions

* [ ] `iroh_sync::EntrySignature` should the signatures include a namespace prefix?
* [ ] do we want the 1:1 mapping of `NamespaceId`and gossip `TopicId`, or would the topic id as a hash be better?

#### Other TODOs collected from the code

* [ ] Port `hammer` and `fs` commands from REPL example to iroh cli
* [ ] docs/naming: settle terminology about keypairs, private/secret/signing keys, public keys/identifiers and make docs and symbols consistent
* [ ] Make `bytes_get` streaming in the RPC interface
* [ ] Allow to unset the subscription on a replica
* [ ] `iroh-sync` shouldn't depend on `iroh-bytes` only for `Hash` type -> #1354
* [ ] * [ ] Move `sync::live::PeerSource` to iroh-net or even better -> #1354
* [ ] `StoreInstance::put` propagate error and verify timestamp is reasonable.
* [ ] `StoreInstance::get_range` implement inverted range
* [ ] `iroh_sync`: Remove some items only used in tests (marked with #[cfg(test)])
* [ ] `iroh_sync` fs store: verify get method fetches all keys with this namespace
* [ ] `ranger::SimpleStore::get_range`: optimize
* [ ] `ranger::Peer` avoid allocs?
* [ ] `fs::StoreInstance::get_fingerprint` optimize
* [ ] `SyncEngine::doc_subscribe` remove unwrap, handle error

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

File: iroh/src/database/flat/writable.rs
Comment (Arqu) [inline iroh/src/database/flat/writable.rs:1]: IIUC this is no longer used and should be dropped.

File: iroh/examples/sync.rs
Comment (flub) [inline iroh/examples/sync.rs:138]: not that this needs to change, but i believe the "async way" is to use a oneshot channel for this.

File: iroh/src/metrics.rs
Comment (flub) [inline iroh/src/metrics.rs:33]: doesn't seem that useful to me.  a histogram of download speeds or something might be more useful.  but just removing this would also be fine i think

File: iroh-gossip/src/net.rs
Comment (Frando) [reply iroh-gossip/src/net.rs:483]: Those were not added in this PR but in #1149. `me` is a field because it is a simple type (32 byte peer id) whereas `event` is a complex type (enum with tuple variantes and potentially a lot of data in them) so I thought debug-printing makes more sense. But also I'm not really sure when to use fields vs debug logging in tracing. I created a separate PR for this as its unrelated to this PR: https://github.com/n0-computer/iroh/pull/1390

File: iroh-sync/src/lib.rs
Comment (flub) [inline iroh-sync/src/lib.rs:4]:
```suggestion
//! [Entrys](Entry). Each entry is identified by a key, its author, and the replica's
```
?

But if cargo-doc is happy then it's fine i guess...

File: iroh-sync/Cargo.toml
Comment (Frando) [reply iroh-sync/Cargo.toml:19]: I don't know, but this is the same as on `main`: https://github.com/n0-computer/iroh/blob/main/iroh/Cargo.toml#L25

File: iroh/Cargo.toml
Comment (Frando) [reply iroh/Cargo.toml:68]: Good point. I think this is a leftover from the initial hacky integration of iroh-sync. Removed, let's see if CI says everything fine.

File: iroh-sync/src/keys.rs
Comment (flub) [inline iroh-sync/src/keys.rs:36]: Seems redundant since `AuthorId::as_bytes` already exists

File: iroh/examples/hello-world.rs
Comment (Frando) [reply iroh/examples/hello-world.rs:25]: Because the `iroh::node::Node` is now generic over the doc store. What we could do is to add a `NoopDocStore` or `DisabledDocStore` that would return errors for all method calls, and make that the default for the generic. WDYT? Or other ideas on how to deal with "we need a generic for this but want a disabled default"?

File: iroh-sync/Cargo.toml
Comment (Frando) [reply iroh-sync/Cargo.toml]: fixe

File: iroh-sync/Cargo.toml
Comment (flub) [reply iroh-sync/Cargo.toml:19]: fair enough :shrug:

File: iroh/Cargo.toml
Comment (Frando) [reply iroh/Cargo.toml]: Removed in the latest commit and CI is as green as it gets :+1:

File: iroh-sync/src/ranger.rs
Comment (flub) [inline iroh-sync/src/ranger.rs:58]: why is this available both directly and via the trait?

File: iroh-sync/src/ranger.rs
Comment (flub) [inline iroh-sync/src/ranger.rs:90]: isn't this `to_fingerprint` since it returns something owned?

File: iroh-gossip/src/proto/util.rs
Comment (Frando) [reply iroh-gossip/src/proto/util.rs]: Good catch. Fixed in the latest commit.

File: iroh-sync/src/ranger.rs
Comment (flub) [inline iroh-sync/src/ranger.rs:575]: do you need to allocate this?  Seems like you only get the `len` from it?

File: iroh/src/download.rs
Comment (Frando) [reply iroh/src/download.rs:42]: I think I agree. The downloader will be overhauled in #1344 so not sure if it's worth changing it here.

File: iroh/src/metrics.rs
Comment (Frando) [reply iroh/src/metrics.rs:33]: This was added because @Arqu said "only do counters in metrics if you can" and @b5 wanted to have download speed stats for the demo. So I'd defer to them on what to do here :)

File: iroh-sync/src/keys.rs
Comment (Frando) [reply iroh-sync/src/keys.rs:36]: Yup I agree, @dignifiedquire added it, are you fine with removing it again?

Comment (Frando) [reply iroh-sync/src/keys.rs]: Removed it

File: iroh-sync/src/ranger.rs
Comment (dignifiedquire) [reply iroh-sync/src/ranger.rs:575]: Unfortunately you can't `count` an iterator without consuming it, so I haven't found a way around it.

File: iroh-sync/src/ranger.rs
Comment (dignifiedquire) [reply iroh-sync/src/ranger.rs:58]: Because they take different arguments, and the standalone version is useful to implement the trait on other types.

File: iroh-net/src/key.rs
Comment (dignifiedquire) [inline iroh-net/src/key.rs:60]: Is this really needed when we have `TryFrom<&[u8>`?

File: iroh-sync/src/keys.rs
Comment (dignifiedquire) [inline iroh-sync/src/keys.rs:150]: you can avoid these manual impls by using `derive_more::Display` much easier to do

File: iroh/src/commands/sync.rs
Comment (dignifiedquire) [inline iroh/src/commands/sync.rs:75]: why is this commented out?

File: iroh/src/commands/sync.rs
Comment (dignifiedquire) [inline iroh/src/commands/sync.rs:124]: should be the same as with get `#[clap(short, long)]` otherwise this is a positional argument

File: iroh/src/commands/sync.rs
Comment (dignifiedquire) [inline iroh/src/commands/sync.rs:151]: this should probably have the option to filter by author?

File: iroh/src/commands/sync.rs
Comment (dignifiedquire) [inline iroh/src/commands/sync.rs:171]: delete?

File: iroh/src/rpc_protocol.rs
Comment (dignifiedquire) [inline iroh/src/rpc_protocol.rs:274]: lets either fill in the `todo`s or remove them

File: iroh/src/node.rs
Comment (dignifiedquire) [inline iroh/src/node.rs:302]: we should do sth about this before merging

File: iroh/src/node.rs
Comment (dignifiedquire) [inline iroh/src/node.rs:313]: you can use a oneshot here

File: iroh/src/node.rs
Comment (dignifiedquire) [inline iroh/src/node.rs:485]: the error from the spawn is never handled, should at least print it

File: iroh-net/src/key.rs
Comment (Frando) [reply iroh-net/src/key.rs:60]: For me having a from_bytes method seems convenient. Eg all the `ed22591_dalek` types have it as well. I can also remove it though..

File: iroh/src/client.rs
Comment (Frando) [reply iroh/src/client.rs]: Done

File: iroh/src/client.rs
Comment (Frando) [reply iroh/src/client.rs]: I added the full impl.

File: iroh/src/commands/sync.rs
Comment (Frando) [reply iroh/src/commands/sync.rs:75]: leftover, removed

File: iroh/src/commands/sync.rs
Comment (Frando) [reply iroh/src/commands/sync.rs:124]: Changed. However also this is all overhauled in #1356

File: iroh/src/commands/sync.rs
Comment (Frando) [reply iroh/src/commands/sync.rs:171]: removed

File: iroh/src/rpc_protocol.rs
Comment (Frando) [reply iroh/src/rpc_protocol.rs]: removed them for now

File: iroh/src/node.rs
Comment (Frando) [reply iroh/src/node.rs:302]: Actually I think we can just remove the TODO:
`init_metrics_collection` calls `iroh_metrics::core::Core::try_init(|reg, metrics| { ... }).ok()`, whereas `Core::try_init` tries to set a `OnceCell`, which will fail for subsequent invocations, which is silenced by the `ok`.
So - not sure how/if we want to do this differently?

File: iroh/src/node.rs
Comment (Frando) [reply iroh/src/node.rs]: logging the error now.

File: iroh/src/commands/sync.rs
Comment (Frando) [reply iroh/src/commands/sync.rs:139]: Added

File: iroh-sync/src/ranger.rs
Comment (dignifiedquire) [reply iroh-sync/src/ranger.rs:90]: maybe

File: iroh-sync/src/keys.rs
Comment (dignifiedquire) [reply iroh-sync/src/keys.rs:150]: https://docs.rs/derive_more/1.0.0-beta.3/derive_more/derive.Display.html

Comment (Frando) [reply iroh-sync/src/keys.rs:150]: OK, can look into this. However I'm not sure if I do like it better to put actual code into the derive attributes, for me that doesn't really look simpler than the impl here.

File: iroh/src/node.rs
Comment (dignifiedquire) [reply iroh/src/node.rs:302]: The only worry I have with this setup, is that we now can not add any metrics in `main.rs` eg, because the first try_init will succeed and follow ups will not, so only the metrics in here will be added. It is probably be fine but want to call it out.

cc @Arqu

File: iroh/src/client.rs
Comment (Frando) [reply iroh/src/client.rs]: I added this. Needed a new RPC method - I chose `DocInfoRequest`, where the `DocInfoResponse` is currently empty, but that would be where we'd add some info on if the document is sync'ed currently etc.
I kept the signature as currently, and added `try_get_doc() -> Result<Option<Doc>>`. Usually you will want to error if the doc doesn't exist, so I think this is easier to use this way.

File: iroh/src/node.rs
Comment (Frando) [reply iroh/src/node.rs]: Well, you would have to add the metrics in `main.rs` before spawning the node. This should be easily doable.
I added a commit that clarifies the current setup with better method names and comments: https://github.com/n0-computer/iroh/pull/1333/commits/503ff23b6439a4787bd4d02e7d605c65cbfd8022
Or do you have another idea on how to proceed with this?

File: iroh-sync/src/keys.rs
Comment (dignifiedquire) [reply iroh-sync/src/keys.rs:150]: not super important

File: iroh/src/node.rs
Comment (dignifiedquire) [reply iroh/src/node.rs]: I think comments about this are good enough for now 👍

Comment (Arqu) [reply iroh/src/node.rs]: Yeah I think having comments is the best option currently. Due to the metrics setup we can only define them once depending where your entrypoint is. The benefit is that you can randomly drop metrics wherever you like without having to pass weirdo references everywhere.

**Review by dignifiedquire — APPROVED** [review]
a long time coming, and finally a big PR that is not from me <3

**Pull request merged**
